src/test/app/Clawback_test.cpp

```diff
@@ -0,0 +1,971 @@
+            env(fset(alice, asfAllowClawback), ter(tecOWNERS));
+            env.close();
+
+            // bob sets trustline to default limit and pays alice back to delete
+            // the trustline
+            env(trust(bob, USD(0), 0));
+            env(pay(bob, alice, USD(10)));
+
+            BEAST_EXPECT(ownerCount(env, alice) == 0);
+            BEAST_EXPECT(ownerCount(env, bob) == 0);
+
+            // alice now is able to set asfAllowClawback
+            env(fset(alice, asfAllowClawback));
+            env.close();
+            env.require(flags(alice, asfAllowClawback));
+
+            BEAST_EXPECT(ownerCount(env, alice) == 0);
+            BEAST_EXPECT(ownerCount(env, bob) == 0);
+        }
+
+        // Test that one cannot enable asfAllowClawback when
+        // featureClawback amendment is disabled
+        {
+            Env env(*this, features - featureClawback);
+
+            Account alice{"alice"};
+
+            env.fund(XRP(1000), alice);
+            env.close();
+
+            env.require(nflags(alice, asfAllowClawback));
+
+            // alice attempts to set asfAllowClawback flag while amendment is
+            // disabled. no error is returned, but the flag remains to be unset.
+            env(fset(alice, asfAllowClawback));
+            env.close();
+            env.require(nflags(alice, asfAllowClawback));
+
+            // now enable clawback amendment
+            env.enableFeature(featureClawback);
+            env.close();
+
+            // asfAllowClawback can be set
+            env(fset(alice, asfAllowClawback));
+            env.close();
+            env.require(flags(alice, asfAllowClawback));
+        }
+    }
+
+    void
+    testValidation(FeatureBitset features)
+    {
+        testcase("Validation");
+        using namespace test::jtx;
+
+        // Test that Clawback tx fails for the following:
+        // 1. when amendment is disabled
+        // 2. when asfAllowClawback flag has not been set
+        {
+            Env env(*this, features - featureClawback);
+
+            Account alice{"alice"};
+            Account bob{"bob"};
+
+            env.fund(XRP(1000), alice, bob);
+            env.close();
+
+            env.require(nflags(alice, asfAllowClawback));
+
+            auto const USD = alice["USD"];
+
+            // alice issues 10 USD to bob
+            env.trust(USD(1000), bob);
+            env(pay(alice, bob, USD(10)));
+            env.close();
+
+            env.require(balance(bob, alice["USD"](10)));
+            env.require(balance(alice, bob["USD"](-10)));
+
+            // clawback fails because amendment is disabled
+            env(claw(alice, bob["USD"](5)), ter(temDISABLED));
+            env.close();
+
+            // now enable clawback amendment
+            env.enableFeature(featureClawback);
+            env.close();
+
+            // clawback fails because asfAllowClawback has not been set
```

Comment:
regarding adding new features to the rippled codebase, why do we need to enable a new amendment and also set the `asfAllowClawback` flag? Aren't both of them redundant?

I'd like to understand why we have this type of design

Reply:
The new amendment is purposed to be voted by the validators. An amendment is required to pass before the feature can be present on the mainnet in the first place.

Now suppose the amendment passes, we have the `asfAllowClawback`  flag that allows someone can set on their _account level_. Without this flag, the account does not have the ability to claw back funds. So, it is individual preference to turn on this flag or not.